Add fullscreen image view with viewerjs controls integration

carlobortolan opened the pull request:
- Closes #163 and  #154
- Adds fullscreen functionality to large images by either pressing the `f` key or the small icon in the bottom corner of each image.

**Preview** 

https://github.com/user-attachments/assets/1da440c6-922e-401d-afda-7ae68dd5ee3b

(ignore the fact that the images seem cut off - that's just because my screen recording did not record the entire screen, but just the browser window)

sambecker commented:
This looks awesome @carlobortolan!

Someone requested something quite similar in #154.

Any thoughts on explicit zoom behavior? Could be via icons or possibly touch controls, which get trapped by the image? 

Let me know what you think.

carlobortolan commented:
> This looks awesome @carlobortolan!
> 
> Someone requested something quite similar in #154.
> 
> Any thoughts on explicit zoom behavior? Could be via icons or possibly touch controls, which get trapped by the image?
> 
> Let me know what you think.

Glad you like it :raised_hands: 

I had a look at #154 and I think it could be an interesting addition (especially as it provides some nice features out of the box). I've just pushed some updates that now let users either:
- Click on the image to use the viewerjs controls, 
- Press `f` or the "toggle fullscreen"-icon to switch directly to presentation mode / fullscreen.

https://github.com/user-attachments/assets/20c81b2c-c492-431a-80f3-bfdfaed9ec13

Tested on Firefox 134.0 (64-bit) and Chromium Version 131.0.6778.264 (Official Build) Arch Linux (64-bit)

sambecker commented:
Thanks for checking that out! Ideally we would add basic zoom/pan behavior without the need for a dependency. Do you have any sense for how much effort that would be?

Relatedly, do you have a link to a deployment I could test out?

carlobortolan commented:
> Thanks for checking that out! Ideally we would add basic zoom/pan behavior without the need for a dependency. Do you have any sense for how much effort that would be?

I initially tried updating the image component or creating a custom viewer component, but was not able to achieve a result as good as the one with jsviewer. 
I think that jsviewer is a relatively easy way to implement this without adding much complexity (in comparison to implementing a custom image viewer), though I'll leave the decision up to you.

> Relatedly, do you have a link to a deployment I could test out?

Yes, I’ve deployed a preview on Vercel. You can check it out here: [https://exif-photo-blog-1-git-main-carlobortolans-projects.vercel.app?_vercel_share=nW7gfuFa3prvZC9Cy0PfEVHu8Na4KMYT](https://exif-photo-blog-1-git-main-carlobortolans-projects.vercel.app/?_vercel_share=nW7gfuFa3prvZC9Cy0PfEVHu8Na4KMYT)

sambecker commented:
Ok, that makes sense. I'm open to using a framework if it makes for a better experience, though I don't think we need to offer much functionality.

Basic requirements:
- be able to zoom in and out with touch controls
- pan around my dragging
- buttons for zooming in, out, and resetting to 100% (ideally these are custom buttons/icons)

A few tweaks/additions:
- I liked the black background in your original build—could be nice to try that again
- think that clicking outside the image should dismiss the viewer
- would be nice if we could  open the viewer whenever we detect a spread touch gesture over an image

WDYT?

carlobortolan commented:
Thanks for clarifying this!
I reintroduced the black background as you suggested and clicking outside the image when using the viewer closes it (see video below & [demo on Vercel](https://exif-photo-blog-1-carlobortolan-carlobortolans-projects.vercel.app)).

https://github.com/user-attachments/assets/f33f705f-c600-42c9-bf3f-14a8b20e7930

Regarding the spread touch gesture to open the viewer, I tested it locally but found it to be a bit too fiddly/tedious on mobile browsers, so I’ve left it out for now, but it can revisited in the future if needed.

So, to summarize: 
- A user can enter a fullscreen mode to zoom in/out and drag the image with touch, mouse and button controls by clicking on an image
- By pressing the `f` key or clicking the icon in the corner of an image, the user can switch to a fullscreen presentation mode (similar to PowerPoint's presentation mode)

If this works for you, we can go ahead and merge it.

sambecker commented:
This is looking great!

A few final thoughts:
- I wonder if we should offer the zoom functionality on the home page—it feels a bit inconsistent offering on large images only when they're details
- I also wonder if this should be offered initially as an ENV VAR customization so we don't surprise people?
- I think the reset button can go back to the "1:1" icon (maybe in our standard font?) and go back to the middle position between the other two buttons

carlobortolan commented:
> I wonder if we should offer the zoom functionality on the home page—it feels a bit inconsistent offering on large images only when they're details

Currently, the new functionality is enabled by default on all PhotoDetailPages. Enabling it on the home page would mean that clicking on an image would open the fullscreen and zoom controls instead of navigating to its DetailPage, making the DetailPage more difficult to find (at least I usually click on the image itself to find out more details).
However, if you prefer, I can change it to also enable the controls for images on the home feed.

> I also wonder if this should be offered initially as an ENV VAR customization so we don't surprise people?

Yes, that's a good idea; I've added a new env variable called `NEXT_PUBLIC_IMAGE_ACTIONS` which defaults to `0` and updated the README accordingly:
![image](https://github.com/user-attachments/assets/902eeb7b-8bce-4626-9f03-4fa0d82f2ec8)

> I think the reset button can go back to the "1:1" icon (maybe in our standard font?) and go back to the middle position between the other two buttons

Viewerjs doesn't have an option yet to change the position of the icons, but I've updated it to use "1:1" in our font:
![image](https://github.com/user-attachments/assets/4e43445b-db1f-4e2c-94f0-b386550bcda4)

sambecker commented:
Did something change regarding image aspect ratios in this latest build? The thumbnails and some of the image mattes look a bit off. Relatedly, should this functionality live in a new abstraction/component vs being commingled with `<ImageWithFallback />` which feels like a primitive that shouldn't know too much about higher level features like full screen zooming/panning, and is used broadly in the application.

carlobortolan commented:
I haven't changed anything in particular regarding the fullscreen and mattes, but it might have to do with the env config of that specific deployment (I'll keep an eye on this though).

To be able to use the image controls, I need to either work directly with the `ImageWithFallback` or the `ImageLarge` component. However, I think I can clean it up a bit; e.g., like this:
```tsx
export default function ImageLarge(props: ImageProps & { enableImageActions?: boolean }) {
    // ...
    <ImageActions enableImageActions={enableImageActions}> {/* <-- new component that integrates viewerjs etc. */}
      <ImageWithFallback {...{
        ...rest,
        blurCompatibilityLevel: blurCompatibilityMode ? 'high' : 'none',
        width: IMAGE_WIDTH_LARGE,
        height: Math.round(IMAGE_WIDTH_LARGE / aspectRatio),
      }} />
    </ImageActions>
  );
}
```

In this case, the `ImageActions` component contains all zoom/fullscreen related functionality, and the `ImageWithFallback` components remain unchanged; would something like this be ok?

sambecker commented:
Nice, that feels a lot better, and I think I'd expect `<ImageActions />` to be rendered from `<PhotoLarge />` instead of `<ImageLarge />`.

carlobortolan commented:
> Nice, that feels a lot better, and I think I'd expect `<ImageActions />` to be rendered from `<PhotoLarge />` instead of `<ImageLarge />`.

Fixed it in my latest commit :+1:

sambecker commented:
This looks great!

If you're good with it, I might merge this into a temporary branch and do some final testing/tweaks before deploying to prod.

Let me know if there's anything else you'd like to contribute.

carlobortolan commented:
Thanks, sounds good to me :+1: 
If I find any other features or bugs I'll let you know.
